library/src/main/java/org/apache/apex/malhar/lib/window/accumulation/AbstractPojoJoin.java

```diff
@@ -121,136 +120,109 @@ private void createSetters()
     int leftStreamIndex = getLeftStreamIndex();
-    List<Map<String, Object>> leftStream = accu.get(leftStreamIndex);
-    List<Map<String, Object>> rightStream = accu.get((leftStreamIndex + 1) % 2);
-    for (Map<String, Object> lMap : leftStream) {
-      boolean gotMatch = false;
-      for (Map<String, Object> rMap : rightStream) {
-        Map<String, Object> tempMap = joinTwoMapsWithKeys(lMap, rMap);
-        if (tempMap != null) {
-          result.add(tempMap);
-          gotMatch = true;
+    Multimap<List<Object>, Object> leftStream = accu.get(leftStreamIndex);
+    Multimap<List<Object>, Object> rightStream = ArrayListMultimap.create(accu.get((leftStreamIndex + 1) % 2));
+    Map<String,PojoUtils.Getter> leftGettersStream = leftStreamIndex == 0 ? gettersStream1 : gettersStream2;
+    Map<String,PojoUtils.Getter> rightGettersStream = leftStreamIndex == 1 ? gettersStream1 : gettersStream2;
+    for (List lMap : leftStream.keySet()) {
+      Collection<Object> left = leftStream.get(lMap);
+      if (rightStream.containsKey(lMap)) {
+        Collection<Object> right = rightStream.get(lMap);
+        Object o;
+        try {
+          o = outClass.newInstance();
+        } catch (Throwable e) {
+          throw Throwables.propagate(e);
         }
-      }
-      if (!gotMatch) {
-        addNonMatchingResult(result, lMap, rightStream.get(0).keySet());
+        for (Object lObj:left) {
+          for (Object rObj:right) {
+            setObjectForResult(leftGettersStream, lObj,o);
+            setObjectForResult(rightGettersStream, rObj,o);
+          }
+          result.add(o);
+        }
+        rightStream.removeAll(lMap);
+      } else {
+        addNonMatchingResult(left, leftGettersStream, result);
       }
     }
+    addNonMatchingRightStream(rightStream, rightGettersStream, result);
     return result;
   }
 
-  public abstract void addNonMatchingResult(List<Map<String, Object>> result, Map<String, Object> requiredMap, Set nullFields);
-
-  public abstract int getLeftStreamIndex();
+  public abstract void addNonMatchingResult(Collection<Object> left, Map<String,PojoUtils.Getter> leftGettersStream, List<Object> result);
 
+  public abstract void addNonMatchingRightStream(Multimap<List<Object>, Object> rightStream, Map<String,PojoUtils.Getter> rightGettersStream, List<Object> result);
 
-  public Map<String, Object> joinTwoMapsWithKeys(Map<String, Object> map1, Map<String, Object> map2)
-  {
-    int lIndex = getLeftStreamIndex();
-    for (int i = 0; i < leftKeys.length; i++) {
-      String key1 = lIndex == 0 ? leftKeys[i] : rightKeys[i];
-      String key2 = lIndex == 1 ? leftKeys[i] : rightKeys[i];
-      if (!map1.get(key1).equals(map2.get(key2))) {
-        return null;
-      }
-    }
-    for (String field : map2.keySet()) {
-      if (!keySetStream2.contains(field)) {
-        map1.put(field, map2.get(field));
-      }
-    }
-    return map1;
-  }
+  public abstract int getLeftStreamIndex();
```

Comment:
Can you add a javadoc comment for this containing, what is expected to happen here and when will this be called?